spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,346 @@
+
+Blockchains also need the ability to introspect their own root-of-trust (with `getRootOfTrust`) in order to confirm that the connecting chain has stored the correct one.
+
+#### Subprotocols
+
+Subprotocols are defined as a set of datagram types and a `handleDatagram` function which must be implemented by the state machine of the implementing blockchain. Datagrams must be relayed between chains by an external process. This process is assumed to behave in an arbitrary manner — no safety properties are dependent on its behavior, although progress is generally dependent on the existence of at least one correct relayer process. Further discussion is deferred to [ICS 18: Relayer Algorithms](../spec/ics-18-relayer-algorithms).
+
+IBC subprotocols are reasoned about as interactions between two chains `A` and `B` — there is no prior distinction between these two chains and they are assumed to be executing the same, correct IBC protocol. `A` is simply by convention the chain which goes first in the subprotocol and `B` the chain which goes second. Protocol definitions should generally avoid including `A` and `B` in variable names to avoid confusion (as the chains themselves do not know whether they are `A` or `B` in the protocol).
+
+This ICS defines four subprotocols: opening handshake, header tracking, closing handshake, and closing by equivocation.
+
+##### Opening Handshake
+
+The opening handshake subprotocol serves to initialize roots of trust for two chains on each other and negotiate an agreeable connection version.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+```golang
+type ConnOpenInit struct {
+  // Identifier to use for connection on chain A
+  Identifier  identifier
+  // Desired identifier to use for connection on chain B
+  Identifier  desiredCounterpartyIdentifier
+  // Desired version for connection
+  Version     desiredVersion
+  // Root-of-trust for chain B
+  RootOfTrust rootOfTrust
+}
+```
+
+```golang
+type ConnOpenTry struct {
+  // Desired identifier to use for connection on chain B
+  Identifier        desiredIdentifier
+  // Identifier for connection on chain A
+  Identifier        counterpartyIdentifier
+  // Desired version for connection
+  Version           desiredVersion
+  // Root-of-trust for chain A
+  RootOfTrust       rootOfTrust
+  // Proof of stored INIT state on chain A
+  AccumulatorProof  proofInit
+}
+```
+
+```golang
+type ConnOpenAck struct {
+  // Identifier for connection on chain A
+  Identifier        identifier
+  // Identifier for connection on chain B
+  Identifier        agreedCounterpartyIdentifier
+  // Agreed version for connection
+  Version           agreedVersion
+  // Proof of stored TRY state on chain B
+  AccumulatorProof  proofTry
+}
+```
+
+```golang
+type ConnOpenConfirm struct {
+  // Identifier for connection on chain B
+  Identifier        identifier
+  // Proof of stored OPEN state on chain A
+  AccumulatorProof  proofAck
+}
+```
+
```

Comment:
Read through all messages a few times an it is not clear to me: a connection is between A and B. Shouldn't the connection identifier be the combination of identifiers on A and B? Also, why is one side proposing an identifier for the other side? Are we trying to solve cases where two chains have the same name?

Reply:
We have only local identifiers, not global ones. I think we want chains to know each other's identifier for themselves, but I don't think we need to require that they have the same identifier (although that could be recommended by convention).